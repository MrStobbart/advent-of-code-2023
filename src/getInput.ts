export const getInputForDay = (day: number) => {
  const fileName = `./data/day${day}.input.txt`;
  try {
    const file = Bun.file(fileName);
    return file.text();
  } catch (error) {
    throw new Error(
      `Could not load file with name ${fileName}`,
      error as Error
    );
  }
};
