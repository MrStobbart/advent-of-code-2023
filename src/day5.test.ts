import { describe, expect, it } from "bun:test";
import {
  getDestination,
  getLowestLocation,
  parseAlamanc,
  prepareMap,
  seedParserPart2,
} from "./day5";
import { getInputForDay } from "./getInput";

const testInput = `
seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
`;

const input = await getInputForDay(5);

describe("Part 1", () => {
  const testMap = `
seed-to-soil map:
50 98 2
52 50 48`;

  const preparedMap = prepareMap(testMap);

  it("prepares a map correctly", () => {
    expect(preparedMap).toEqual({
      name: "seed-to-soil",
      ranges: [
        { start: 98, end: 99, offset: -48 },
        { start: 50, end: 97, offset: 2 },
      ],
    });
  });

  it.each([
    { source: 79, destination: 81 },
    { source: 14, destination: 14 },
    { source: 55, destination: 57 },
    { source: 13, destination: 13 },
  ])("gets the correc destination", ({ source, destination }) => {
    expect(getDestination(preparedMap, source)).toBe(destination);
  });

  it("parses an alamanc correctly", () => {
    expect(parseAlamanc(testInput)).toMatchSnapshot();
  });

  it("gets the correct lowest location", () => {
    const alamanc = parseAlamanc(testInput);
    expect(getLowestLocation(alamanc)).toBe(35);
  });

  it("calculates the correct result", () => {
    const alamanc = parseAlamanc(input);
    expect(getLowestLocation(alamanc)).toBe(331445006);
  });
});

describe("Part 2", () => {
  it("calculates the lowest location with seed ranges", () => {
    const alamanc = parseAlamanc(testInput, seedParserPart2);
    expect(getLowestLocation(alamanc)).toBe(46);
  });

  it("calculates the correct result", () => {
    // const alamanc = parseAlamanc(input, seedParserPart2);
    console.log(alamanc);

    // expect(getLowestLocation(alamanc)).toBe(46);
    expect(35).toBe(46);
  });
});
