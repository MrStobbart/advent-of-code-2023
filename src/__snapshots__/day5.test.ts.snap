// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Part 1 parses an alamanc correctly 1`] = `
{
  "maps": [
    {
      "name": "seed-to-soil",
      "ranges": [
        {
          "end": 99,
          "offset": -48,
          "start": 98,
        },
        {
          "end": 97,
          "offset": 2,
          "start": 50,
        },
      ],
    },
    {
      "name": "soil-to-fertilizer",
      "ranges": [
        {
          "end": 51,
          "offset": -15,
          "start": 15,
        },
        {
          "end": 53,
          "offset": -15,
          "start": 52,
        },
        {
          "end": 14,
          "offset": 39,
          "start": 0,
        },
      ],
    },
    {
      "name": "fertilizer-to-water",
      "ranges": [
        {
          "end": 60,
          "offset": -4,
          "start": 53,
        },
        {
          "end": 52,
          "offset": -11,
          "start": 11,
        },
        {
          "end": 6,
          "offset": 42,
          "start": 0,
        },
        {
          "end": 10,
          "offset": 50,
          "start": 7,
        },
      ],
    },
    {
      "name": "water-to-light",
      "ranges": [
        {
          "end": 24,
          "offset": 70,
          "start": 18,
        },
        {
          "end": 94,
          "offset": -7,
          "start": 25,
        },
      ],
    },
    {
      "name": "light-to-temperature",
      "ranges": [
        {
          "end": 99,
          "offset": -32,
          "start": 77,
        },
        {
          "end": 63,
          "offset": 36,
          "start": 45,
        },
        {
          "end": 76,
          "offset": 4,
          "start": 64,
        },
      ],
    },
    {
      "name": "temperature-to-humidity",
      "ranges": [
        {
          "end": 69,
          "offset": -69,
          "start": 69,
        },
        {
          "end": 68,
          "offset": 1,
          "start": 0,
        },
      ],
    },
    {
      "name": "humidity-to-location",
      "ranges": [
        {
          "end": 92,
          "offset": 4,
          "start": 56,
        },
        {
          "end": 96,
          "offset": -37,
          "start": 93,
        },
      ],
    },
  ],
  "seeds": [
    {
      "end": 79,
      "start": 79,
    },
    {
      "end": 14,
      "start": 14,
    },
    {
      "end": 55,
      "start": 55,
    },
    {
      "end": 13,
      "start": 13,
    },
  ],
}
`;
